### PR TITLE
Initial parser log proto

### PR DESCRIPTION
**Pull request by AlexanderKapelyukhovskiy**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -1,9 +1,12 @@
-import { log } from './utils/console-utils'
+import {parser} from './log-parser'
+import {builder} from './log-tree-builder'
+import LogEntry from 'log-entry'
 
+let head = new LogEntry('00:00:00', 'HEAD of the Log', 0) //TODO: parse date from Log
```
Comment (Restuta) [inline lib/index.js]: easy with momentjs or vanilla `new Date()`

File: lib/log-entry.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export default function(
+    logDate, logMessage, level, parent, leaves) {
+  this.logDate = logDate
+  this.logMessage = logMessage
+  this.level = level
+  this.parent = parent
```
Comment (Restuta) [inline lib/log-entry.js]: apparently "parent" is a reserved variable on the client, while it's a server-side code I'd recommend to use `parentNode`

File: lib/log-parser.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+const parseStartDate = (lineWithDate) => {
+  const result = lineWithDate.match(startDatePattern)
+  if (result) {
+    return moment(result[1], 'ddd MMM DD HH:mm:ss PST yyyy').toDate()
+  }
+  return null
+}
+
+const parseLogDate = (logDateString, startDate) => {
```
Comment (Restuta) [inline lib/log-parser.js:16]: why this can't be done with moment?

File: lib/log-parser.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  if (logDate < startDate) {
+    //if log finieshed next day
+    logDate.setDate(logDate.getDate() + 1)
+  }
+
+  return logDate
+}
+
+const parseLine = (logLine, startDate) => {
+  const result = logLine.match(logEntryPattern)
```
Comment (Restuta) [inline lib/log-parser.js]: name of the variable is meaningless, result of what? It's actually an array of matches, so `logEntryMatches` is far better name

File: lib/log-parser.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
+
+import fs from 'fs'
+import readline from 'readline'
+import LogEntry from 'log-entry'
+import moment from 'moment'
+
+const startDatePattern = /^Started '(.*)' on(.*)$/
+const logEntryPattern = /^\[(\d{2}:\d{2}:\d{2})\][ |\w]:(\t*)(.*)$/
```
Comment (Restuta) [inline lib/log-parser.js:6]: Triple checked that it maches all log entries?

File: lib/log-parser.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    if (tabs && tabs.length > 0) {
+      level = tabs.length
+    }
+    return new LogEntry(logDate, message, level)
+  }
+
+  return null
+}
+
+const parser = (filePath, callBack) => {
```
Comment (Restuta) [inline lib/log-parser.js]: collbacks are so 2000, use promises or generators or even async/await, also `callback` is always all lowercase or just `cb` in node world

File: lib/log-parser.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  const result = logLine.match(logEntryPattern)
+
+  if (result) {
+    const logDate = parseLogDate(result[1], startDate);
+    const tabs = result[2]
+    const message = result[3]
+    let level = 0
+
+    if (tabs && tabs.length > 0) {
+      level = tabs.length
```
Comment (Restuta) [inline lib/log-parser.js:44]: Have you double checked that it's tabs and not spaces?

File: lib/index.js [hunk 8]
```diff
@@ -1,9 +1,12 @@
...
 
-log.json({
-  hello: 'world'
-})
-
+parser(
```
Comment (Restuta) [inline lib/index.js]: this is confusing name, a function should be a verb, it should be either `logParser.parse` or just `parseLog`

**Review by Restuta — CHANGES_REQUESTED** [review]

**Comment by Restuta** [comment]
Those comments is one thing, but I have a bigger concern with the current approach, since most interesting part is to assign to each node a numerical value, time it took in seconds to execute this line. TeamCity puts into log times sometimes, I am not sure how reliable is that and if we can trust it.

If we _can't_ trust it, then I think we can calculate time it took in the following way:

Let's say we are reading following log:

```
[1m10s] Entry A
  [1m12s] Entry B
  [1m15s] Entry C
```

correct tree would look like:
```
     A(2+3=5s)
  /      \
B(2s)  C(3s)
```

But in order to calculate that we need to do two things: 
1. read lines in pairs and subtract Entry B time from Entry A time
2. after tree is built we should have all leaf values set, so we should need to traverse the tree in a special way updating leaf parents to the sum of the leafs. *This is based under assumption that time it took to execute a parent = sum of all children times*, which may not be true. If it's not true, then we have option 3.
3. More reliable way to calculate times could be to just have a sorted tree, where on each level tree is sorted in the order log is sorted (by time), so then after tree is built we can do breadth-first traversal and calculate time it took by just doing node (C - B), *This is based under assumption that time it took to execute a parent NOT NECESSARILY equals to sum of all children times, in this case we can even add a "pseudo-node" with the delta if it's non 0, so on the flame-chart we can see time not reported by teamcity (spent on something else), I'd assume it's pretty minimal and maybe we can ignore that for now.

File: lib/log-parser.js (hunk 5 above)
Comment (AlexanderKapelyukhovskiy) [reply lib/log-parser.js:6]: @Restuta, could you please clarify this note?

File: lib/log-parser.js (hunk 3 above)
Comment (AlexanderKapelyukhovskiy) [reply lib/log-parser.js:16]: We probably can simplify this part of parsing. Right now we have only time inside log entries like 
```
[02:30:07] : Finalize build settings
```

and we also have full start date in second line of log:
```
Started 'Thu Dec 15 02:37:22 PST 2016' on 'SAC-CORP-TC2' by 'Schedule Trigger
```

so we need to get time from second log line and using time from log entry create full log date for each log entry

File: lib/log-parser.js (hunk 7 above)
Comment (AlexanderKapelyukhovskiy) [reply lib/log-parser.js:44]: yes
Alt

**Review by Restuta — APPROVED** [review]

**Comment by Restuta** [comment]
What do you think about my comment?

File: index.html [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    <meta charset="utf-8">
+    <link rel="stylesheet" type="text/css" href="http://spiermar.github.io/d3-flame-graph/d3.flameGraph.css">
+  </head>
+  <body>
+    <div class="container">
+      <div id="chart">
+      </div>
+    </div>
+
+    <script type="text/javascript" src="http://d3js.org/d3.v3.min.js"></script>
```
Comment (Restuta) [inline index.html]: That's an old version of D3, we need to use v4

**Review by Restuta — CHANGES_REQUESTED** [review]

File: index.html (hunk 9 above)
Comment (AlexanderKapelyukhovskiy) [reply index.html]: I was not able to find flame graph library integrated with d3 v4. 
As you can see from https://github.com/spiermar/d3-flame-graph/blob/master/package.json
it depends on "d3": "^3.5.6"

**Review by Restuta — APPROVED** [review]

File: lib/visualizable-log-tree-builder.js [hunk 10]
```diff
@@ -15,13 +15,42 @@ const getNextLogDate = (node) => {
   }
 }
 
+const mmss = secs => {
```
Comment (Restuta) [inline lib/visualizable-log-tree-builder.js]: wins a :1st_place_medal:  for easy to understand name

File: lib/visualizable-log-tree-builder.js [hunk 11]
```diff
@@ -15,13 +15,42 @@ const getNextLogDate = (node) => {
   }
 }
 
+const mmss = secs => {
+  if(secs <= 0)
```
Comment (Restuta) [inline lib/visualizable-log-tree-builder.js]: always use `{`, need to ESLint this

**Pull request merged**